### ISSUE TITLE
GitOps

### ISSUE THREAD

**Issue by rusenask** [keel-hq/keel#220]
There were quite a few people asking on various occasions regarding GitOps support in Keel. There's a nice article here https://www.weave.works/blog/gitops-operations-by-pull-request that describes this model of approving/making changes via Git operations. 

It might be worth it to consider adding something similar to Keel. Even though when building Keel I had a similar pipeline that drives changes via Git operations (tag a release on github which results in image build and then auto deploy), I never though about having another "gate keeper" that blocks before the actual deployment update. That's when we added approvals that can be used via Slack, HipChat or webhooks.

## Proposal A

A new proposal (I would like to encourage suggesting different models) would be to add a new way of approving changes via Git:

1. Keel would have a new CRD that stores information about remote repositories for GitOps
2. Each repository holds either whole deployment/statefulset/helm chart/daemonset or just specific configuration. I would lean to specific config file here just to make it simpler and more dynamic. In the end we only care about image repo with tag, namespace and resource name. 
3. Once new image is available **and GitOps config has an older version**, Keel creates a PR or a new branch that a user has to manually merge.
4. Once merge is done we trigger an event to providers (Helm, Kubernetes) to re-run updates and this time it will succeed because GitOps config already has a newer version.

It would work quite similarly to our current approvals (https://keel.sh/v1/guide/documentation.html#Approvals) but instead of sending notification to Slack, it would create a new branch and require it to be merged.

## Proposal B

Same as proposal A but instead of making a new branch with a change, just passively sync repository and once a new config is available, send a trigger event to providers to start an update. This would require even less work.

Both proposals seem easy to use and would provide a good update history via git. Any downgrades in versions could be accompanied by `force` policy to enforce updating to a lower version. However, they would still require quite a bit of plumbing to add a CRD for remote git repo management and working with git. 

Suggestions needed! :)

**Comment by marcuslindfeldt** [comment]
I think there are use cases for both proposals. E.g if I just wan't continuous deployment to my dev,test,staging environments I would suggest that keel simply makes commits to the correct files in the repo,branch specified in the CRD.

But for continuous delivery to production a pull request based flow would be more suitable based on some condition like semver tagged images or something.

Regarding rollbacks/downgrades I would suggest a "roll-forward" approach instead.  It would be great to simply `git revert <commitsha>` followed by a `git push`. Then the webhook would trigger an event and keel would deploy the new version.

As I understand GitOps it's important that git is the source of truth. So maybe we also need to poll the repo regularly to compare state.

Maybe we can use annotations/labels to identify which files/manifests should be gitops managed?

**Comment by rusenask** [comment]
I think the repo should be separate for gitops operations. For example
if product repo is `github.com/keel-hq/keel` then gitops could be `github.com/keel-hq/keel-ops` or something similar. In that weaveworks talk they made a really good point for this - changes to your repo shouldn't trigger a process again. So for example if you already tagged a release 0.5.0 in your product repo, no additional commits should be needed there. So a separate repo would then get a new commit with that 0.5.0 tag. 

Then, since it's separate, we can control it in any way we want. We could have a directory structure like:
`<namespace>/<kind - deployment/daemonset/statefulset/helm; optional but could be useful>/<name>`

And the file itself could either be a JSON or yaml that stores information that we care about. Another option would be somehow get the deployment.yaml or daemonset from the k8s API (haven't tried doing it) as files so users can edit and we would read/apply them. 

Or we could also start with one and then later add support for both. CRD could have a setting for this behaviour. I would want to experiment with the following setup:

* CRD that has: 
  - remote repo address
  - remote repo branch (so you could have same repo for dev/test/prod)
  - SSH key 
  - namespace/kind/name (how to target our deployment)
* Git control package to push/pull/commit
* GitOps config that has the configuration file with:
  - namespace/kind/name
  - image name
  - image tag
  - auto update true/false (to use policy to decide whether to auto commit new version)
* Sync package that uses GitOps package, Git control package and acts as an approvals gateway. Ideally this would require very little changes to triggers and providers. It would work as a new approvals manager implementation and would fire force update events to providers once git config is changed. 

`git revert` should work as it would just get the current repo state and use it :) We should initially go with just pulling the remote but similar webhook triggers to what we already have for the registries could be later added. 

Does this sound reasonable?

**Comment by Globegitter** [comment]
@rusenask just found out about keel and this proposal sounds pretty good to me. If that was available today like that I would certainly give it a spin :)

**Comment by rusenask** [comment]
@Globegitter no, gitops is not available yet :) I started working on it, just the PoC for now and will update this thread once I have something useful. 

It's going to be quite different than the current main functionality so almost two services in a single container

**Comment by zbindenren** [comment]
It would be great if instead of polling the git repository there would also be an option with webhooks. Github and Gitlab both support webhooks.

**Comment by rusenask** [comment]
Yeah, webhooks might be the first implementation tbh as it's a lot easier, reliable and efficient way of doing it.

**Comment by shinebayar-g** [comment]
Can someone implement Gitlab webhook into Keel?

**Comment by rusenask** [comment]
Hi, please create a new issue with a link to what kind of webhooks you want to have there :) (this one https://docs.gitlab.com/ee/user/project/container_registry.html ?)

**Comment by shinebayar-g** [comment]
Yes sir, I was about to open Issue here. I'll go into more details there.

**Comment by Multiply** [comment]
@rusenask Has there been further developments on this?

We're currently looking at both Flux and Keel, and this feature seems to be the deciding factor for our choice of tooling.

We'd also like to contribute with PRs/testing if that's possible.

**Comment by rusenask** [comment]
Hi, no. After several horror scenarios in another environment where we had gitops I am not so sure if this something that would be nice to have. Flux kept recreating workloads even when "automation" was turned off. That was quite unfortunate as it was recreating etcd and one other stateful set that was relying on etcd. That would always result in a crash (we needed to deploy etcd, then once it's ready restore backup into it and only deploy the 2nd stateful set service). After that fun event I had to remove it from the environment. 

So I think the pros of gitops are:
- might be useful to auto update some config files after git push

cons:
- when something goes wrong then it really ruins your day as it won't fix the situation and might just make it worse
- you no longer can really use your kubectl to quickly scale down/redeploy workloads
- you no longer control your cluster with kubectl in general
- if ordering is required in recreating environment, it can be problematic

So, to sum up. It's good to store your manifests in a git repo and create them all by just `kubectl apply -f gke/` (and update if any config has changed). However, compared to image tags, manifests are usually quite static. 

I guess you just need to evaluate your environment. If you have lots of config files that are going to be changing often, try going with flux and see whether it works for you :)

**Comment by Multiply** [comment]
For us it's more about being able to spin up an identical cluster, if something were to break down.
Without having the versions for each deployment checked into some source of truth (git for example), we can't easily recover our previous state.

Assuming we don't hit the problems you've seen with Flux, all we need to do is monitor specific labels in our docker registry, and it'll handle git for us, allowing us to easily recover any incidents. (Sidenote: I'd be really interested in hearing more of your war stories with GitOps if you have the time)

I don't think I'd trust automation for statefulsets in any case, as I've seen many things go horribly wrong in that department. For that, we'll stick to manually updating deployments using helm or helmfile directly, until we trust the tooling more.

With that said, we don't really need specific order of operations, as all our microservices work on their own, aside from our monolith.

**Comment by severity1** [comment]
i think the solution here is make it so that upon initial deployment `image_tag` will be overridden by keel labels/annotations.

`you're deployment wants to use floop:0.0.8 but your keel policy will enforce floop:0.0.12`

not sure if keel can override before pod is created tho.

**Comment by stephen-dahl** [comment]
@severity1 that is possible with kubernetes mutating webhooks

**Comment by yosbeleg89** [comment]
Is this something that is planned to be implemented? We were using Keel for a while now, we have a git repo with Helm charts that gets applied with Helm when a PR is merged. Now, we upgraded to Helm 3, and every time a PR is merged Helm 3 rollbacks all deployment images updated by keel becasue they are in a previous version and Keel upgrade them again. If something like this would be implemented it would be great since it would solve our issue.

**Comment by jlewi** [comment]
I stumbled on this issue while looking for tooling that could open up PRs to modify Kubernetes manifests in response to upstream changes; e.g. detect when a new image is published to a docker registry; open up a PR to change the image in a deployment to use that.

It looks like keel.sh directly applies the changes to the Kubernetes cluster as opposed to opening up PRs.
I found this fork: https://github.com/Alwinius/bow  which I think opens up PRs.

Following a shift left philosophy, I think you can separate automating updating the source of truth in Git with how you roll it out. So you could use GitOps tools like ArgoCD/Flux or you could manually do `kubectl apply`. Either way its beneficial to have source of truth checked into git and automate keeping the source of truth up to date.

**Comment by MPV** [comment]
FYI: 

It seems `dependabot` (GitHub) has recently started work on support for making PRs on updated helm charts:

https://github.com/dependabot/dependabot-core/issues/2237

And while `version-checker` doesn't support creating PRs like Keel, it can help detecting (and alerting via Prometheus-based alerts) whenever newer images have been released:

https://github.com/jetstack/version-checker
